src/activities/quizzes/ipRestrictions/QuizIpRestrictionsEntity.js

```diff
@@ -0,0 +1,101 @@
+import { Entity } from '../../../es6/Entity';
+import { Actions, Classes } from '../../../hypermedia-constants';
+import { performSirenAction } from '../../../es6/SirenAction.js';
+
+const IP_START_FIELD = 'start';
+const IP_END_FIELD = 'end';
+
+/**
+ * Quiz IP restrictions subentity of a d2l Quiz.
+ */
+export class QuizIpRestrictionsEntity extends Entity {
+
+	/**
+	 * @returns {bool} Whether or not the edit IP restrictions type action is present
+	 */
+
+	canEditIpRestrictions() {
+		return this._entity && this._entity.hasActionByName(Actions.quizzes.ipRestrictions.add);
+	}
+
+	/**
+	 * @returns {object} Returns IP restriction sub entity
+	 */
+
+	getIpRestrictions() {
+		if (!this._entity) {
+			return;
+		}
+
+		const ipRestrictionEntities = this._entity.getSubEntitiesByClass(Classes.quizzes.ip.range);
+
+		if (!ipRestrictionEntities || !ipRestrictionEntities.length) {
+			return [{ start: '', end: '' }];
+		}
+
+		return ipRestrictionEntities.map(({ properties }) => {
+
+			const { start, end, oldStart, oldEnd } = properties;
```

Comment:
This observation is going to affect how you handle this data in the MobX state, and I have some other comments coming about that, but I wasn't intending that we include the `old` values as properties. These are effectively hidden hypermedia state in the field definition that is put there just so that the server knows how to process the action. The client shouldn't be aware it even exists, which is why the `old` fields should be defined as `type="hidden"` in the action. The siren action implementation will automatically include those hidden fields when the client executes that action without the client having to deal with them at all.

This does have other implications for how you are managing the state and how you associate what you are managing in the MobX state with what is represented in the original entity, as you need some way to related the MobX state back to the relevant entity in the original siren response. 

I hadn't fully thought this out, but just thinking about it now, I was wondering if we could use the index of each existing IP restriction in the siren entity as a kind of identifier, that the MobX uses to track which bit of state it is showing in the UI is associated with which entity in the Siren response. This wouldn't be used as an offset based index in the MobX state, because it would be possible to delete an entry in the UI such that what is at position 2 in the UI might be associated with a siren entity at position 1 for example.

e.g. when you return the list of IP restrictions to the MobX state you would return something that looked more like this:

[
{ "start": "192.168.1.1", "end": "192.168.1.10", "id": 0},
{ "start": "192.168.1.20", "end": "192.168.1.30", "id": 10}
]

etc.

New IP restrictions that get added wouldn't have an id in the MobX state which is how you can track that they are new.

I think this would affect the implementation of add and delete below, and we'd probably also want an update that utilized the id field as a property to the update method, so it could find the relevant sub entity to update. The sub entity update action would include the `old` hidden fields, and so no need to be passing them in the `update` method call.